### PR TITLE
Best effort cleanup of staged upload files (#4778)

### LINKED ISSUE #4778: Automatically Cleanup LocalFileSystem Temporary Files
**Is your feature request related to a problem or challenge? Please describe what you are trying to do.**

LocalFileSystem stages uploads to temporary paths ending with `#\d+`. These are ignored by all APIs exposed by LocalFileSystem, however, will still be visible to systems that interact with the filesystem directly, and will continue to consume disk space.

In most cases these will be cleaned up automatically, however, in the event of a crash or an aborted future, these files will be left around

**Describe the solution you'd like**

Some mechanism to automatically reap these files. In the abscence of some asynchronous drop mechanism for futures, it is unclear how this would work, but creating the ticket to track.

**Describe alternatives you've considered**

**Additional context**

### PR DESCRIPTION
**Pull request by tustvold**
# Which issue does this PR close?

Closes #4778

# Rationale for this change
 

# What changes are included in this PR?

# Are there any user-facing changes?

### CODE REVIEW

File: object_store/src/local.rs [hunk 1]
```diff
@@ -274,13 +274,15 @@ impl ObjectStore for LocalFileSystem {
...
-
             file.write_all(&bytes)
-                .context(UnableToCopyDataToFileSnafu)?;
-
-            std::fs::rename(staging_path, path).context(UnableToRenameFileSnafu)?;
-
-            Ok(())
+                .context(UnableToCopyDataToFileSnafu)
+                .and_then(|_| {
+                    std::fs::rename(&staging_path, &path).context(UnableToRenameFileSnafu)
```
Comment (tustvold) [inline object_store/src/local.rs:280]: I'm not really sure how best to test this...

File: object_store/src/local.rs [hunk 2]
```diff
@@ -318,7 +322,6 @@ impl ObjectStore for LocalFileSystem {
         &self,
         location: &Path,
     ) -> Result<Box<dyn AsyncWrite + Unpin + Send>> {
-        #[cfg(not(target_arch = "wasm32"))]
```
Comment (tustvold) [inline object_store/src/local.rs:321]: The entire module is feature gated so this was redundant

File: object_store/src/local.rs [hunk 3]
```diff
@@ -1560,6 +1561,25 @@ mod not_wasm_tests {
...
+
+        let location = Path::from("some_file");
+        let (_, mut writer) = integration.put_multipart(&location).await.unwrap();
+        writer.write_all(b"hello").await.unwrap();
+
+        let file_count = std::fs::read_dir(root.path()).unwrap().count();
+        assert_eq!(file_count, 1);
+        drop(writer);
+
+        tokio::time::sleep(Duration::from_millis(1)).await;
```
Comment (alamb) [inline object_store/src/local.rs:1578]: if we find that this fails intermittently, the other classic pattern is to add a poll loop with maximum wait time

File: object_store/src/local.rs [hunk 4]
```diff
@@ -800,14 +788,14 @@ impl AsyncWrite for LocalUpload {
     fn poll_flush(
         self: Pin<&mut Self>,
         _cx: &mut std::task::Context<'_>,
-    ) -> std::task::Poll<Result<(), io::Error>> {
```
Comment (alamb) [inline object_store/src/local.rs:803]: most of these changes are unrelated, driveby cleanups, right?

**Review by alamb — APPROVED** [review]
Looks good to me. Nice work @tustvold  Thank you
